databroker/src/grpc/kuksa_val_v2/val.rs

```diff
@@ -220,30 +265,139 @@ impl proto::val_server::Val for broker::DataBroker {
+    //            the data type of the addressed signal
+    //       - if the requested value is not accepted,
+    //            e.g. if sending an unsupported enum value
     async fn actuate(
         &self,
-        _request: tonic::Request<proto::ActuateRequest>,
+        request: tonic::Request<proto::ActuateRequest>,
     ) -> Result<tonic::Response<proto::ActuateResponse>, tonic::Status> {
-        Err(tonic::Status::new(
-            tonic::Code::Unimplemented,
-            "Unimplemented",
-        ))
+        debug!(?request);
+        let permissions = match request.extensions().get::<Permissions>() {
+            Some(permissions) => {
+                debug!(?permissions);
+                permissions.clone()
+            }
+            None => return Err(tonic::Status::unauthenticated("Unauthenticated")),
+        };
+        let broker = self.authorized_access(&permissions);
+
+        let actuator_request = request.into_inner();
+        if let Some(value) = actuator_request.value {
+            let opt_signal_id = actuator_request.signal_id.clone();
+            match opt_signal_id {
+                Some(signal_id) => {
+                    match signal_id.signal {
+                        Some(proto::signal_id::Signal::Path(vss_path)) => {
+                            if let Some(id) = broker.get_id_by_path(&vss_path).await {
+                                let result = broker.actuate(&id, &DataValue::from(value)).await;
+                                match result {
+                                    Ok(_) => return Ok(tonic::Response::new(ActuateResponse {})),
+                                    Err(error) => return Err(error.0.to_tonic_status(error.1)),
+                                };
+                            }
+                            return Err(tonic::Status::not_found(format!(
+                                "Invalid vss_path provided {}",
+                                vss_path
+                            )));
+                        }
+                        Some(proto::signal_id::Signal::Id(vss_id)) => {
+                            let result = broker.actuate(&vss_id, &DataValue::from(value)).await;
+                            match result {
+                                Ok(_) => return Ok(tonic::Response::new(ActuateResponse {})),
+                                Err(error) => return Err(error.0.to_tonic_status(error.1)),
+                            };
+                        }
+                        None => {
+                            return Err(tonic::Status::invalid_argument(
+                                "Signal needs to provide Path or Id",
+                            ))
+                        }
+                    };
+                }
+                None => return Err(tonic::Status::invalid_argument("No Signal_Id provided")),
+            };
+        };
+        return Err(tonic::Status::invalid_argument(
```

Comment:
This is totally optional, but if you want less nesting there's quite a few places where you can replace a `match` with a combination of `ok_or(err)` and `?`.
```suggestion
                debug!(?request);
        let permissions = request
            .extensions()
            .get::<Permissions>()
            .ok_or(tonic::Status::unauthenticated("Unauthenticated"))?
            .clone();
        debug!(?permissions);
        let broker = self.authorized_access(&permissions);

        let actuator_request = request.into_inner();
        let value = actuator_request
            .value
            .ok_or(tonic::Status::invalid_argument(
                "Invalid Actuator Request provided",
            ))?;

        let signal = actuator_request
            .signal_id
            .ok_or(tonic::Status::invalid_argument("No signal_id provided"))?
            .signal;

        match &signal {
            Some(proto::signal_id::Signal::Path(path)) => {
                let id = broker
                    .get_id_by_path(path)
                    .await
                    .ok_or(tonic::Status::not_found(format!(
                        "Invalid vss_path provided {}",
                        path
                    )))?;

                match broker.actuate(&id, &DataValue::from(value)).await {
                    Ok(()) => Ok(tonic::Response::new(ActuateResponse {})),
                    Err(error) => Err(error.0.to_tonic_status(error.1)),
                }
            }
            Some(proto::signal_id::Signal::Id(id)) => {
                match broker.actuate(id, &DataValue::from(value)).await {
                    Ok(()) => Ok(tonic::Response::new(ActuateResponse {})),
                    Err(error) => Err(error.0.to_tonic_status(error.1)),
                }
            }
            None => Err(tonic::Status::invalid_argument(
                "Signal needs to provide Path or Id",
            )),
        }
```

Reply:
Thanks for the suggestion! As I'm still a newbie with Rust I'm in particular interested in these kinda "shortcuts" which make code look better and reduce complexity and nesting - so if you have further just keep 'em coming :) I also had the feeling that all of these match-cascades couldn't be the only way to handle it, but I just didn't knew better